src/rpc/blockchain.cpp

```diff
@@ -1327,6 +1334,31 @@ UniValue z_gettreestate(const UniValue& params, bool fHelp)
         res.pushKV("sapling", sapling_result);
     }
 
+    // orchard
+    {
+        UniValue orchard_result(UniValue::VOBJ);
+        UniValue orchard_commitments(UniValue::VOBJ);
+        orchard_commitments.pushKV("finalRoot", pindex->hashFinalOrchardRoot.GetHex());
+        bool need_skiphash = false;
+        OrchardMerkleFrontier tree;
+        if (pcoinsTip->GetOrchardAnchorAt(pindex->hashFinalOrchardRoot, tree)) {
+            CDataStream s(SER_NETWORK, PROTOCOL_VERSION);
+            s << tree;
+            orchard_commitments.pushKV("finalState", HexStr(s.begin(), s.end()));
+        } else {
+            // Set skipHash to the most recent block that has a finalState.
+            const CBlockIndex* pindex_skip = pindex->pprev;
+            while (pindex_skip && !pcoinsTip->GetOrchardAnchorAt(pindex_skip->hashFinalOrchardRoot, tree)) {
+                pindex_skip = pindex_skip->pprev;
+            }
+            if (pindex_skip) {
+                orchard_result.pushKV("skipHash", pindex_skip->GetBlockHash().GetHex());
+            }
+        }
+        orchard_result.pushKV("commitments", orchard_commitments);
+        res.pushKV("orchard", orchard_result);
+    }
```

Comment:
This looks correct, but the Sprout, Sapling, and Orchard code are near-duplicates. The only differences are in the type of the Merkle tree or frontier (which could be a template argument of a helper function), and the expression `pindex->hashFinalOrchardRoot` and member function `pcoinsTip->GetOrchardAnchorAt` (which could be passed in to that function). I will open an issue.